src/Common/Thermodynamics/relations.jl

```diff
@@ -1353,6 +1378,96 @@ function saturation_adjustment(
     end
 end
 
+"""
+    saturation_adjustment_ρpq(
+        param_set,
+        ρ,
+        p,
+        q_tot,
+        phase_type,
+        maxiter,
+        temperature_tol
+    )
+Compute the temperature that is consistent with
+ - `param_set` an `AbstractParameterSet`, see the [`Thermodynamics`](@ref) for more details
+ - `ρ` (moist-)air density
+ - `p` pressure
+ - `q_tot` total specific humidity
+ - `phase_type` a thermodynamic state type
+ - `maxiter` maximum iterations for non-linear equation solve
+ - `temperature_tol` temperature tolerance
+by finding the root of
+
+```
+T - air_temperature_from_ideal_gas_law(
+        param_set,
+        p,
+        ρ,
+        PhasePartition_equil(param_set, T, ρ, q_tot, phase_type),
+    )
+```
+using Newtons method using ForwardDiff.
+See also [`saturation_adjustment`](@ref).
+"""
+function saturation_adjustment_ρpq(
+    param_set::APS,
+    ρ::FT,
+    p::FT,
+    q_tot::FT,
+    phase_type::Type{<:PhaseEquil},
+    maxiter::Int,
+    temperature_tol::FT = eps(FT),
+) where {FT <: Real}
+    tol = SolutionTolerance(temperature_tol)
+    q_pt = PhasePartition(q_tot)
+    T_0 = air_temperature_from_ideal_gas_law(param_set, p, ρ, q_pt)
+    # Use `oftype` to preserve diagonalized type signatures:
+    sol = find_zero(
+        T ->
+            T - air_temperature_from_ideal_gas_law(
+                param_set,
+                oftype(T, p),
+                oftype(T, ρ),
+                PhasePartition_equil(
+                    param_set,
+                    T,
+                    oftype(T, ρ),
+                    oftype(T, q_tot),
+                    phase_type,
+                ),
+            ),
+        NewtonsMethodAD(T_0),
+        CompactSolution(),
+        tol,
+        maxiter,
+    )
+    if !sol.converged
+        if print_warning()
+            @print("-----------------------------------------\n")
+            @print("maxiter reached in saturation_adjustment_ρpq:\n")
+            @print(
+                "    ρ=",
+                ρ,
+                ", p=",
+                p,
+                ", q_tot=",
+                q_tot,
+                ", T = ",
+                sol.root,
+                ", maxiter=",
+                maxiter,
+                ", tol=",
+                tol.tol,
+                "\n"
+            )
+        end
+        if error_on_non_convergence()
+            error("Failed to converge with printed set of inputs.")
+        end
+    end
+    return sol.root
+end
+
```

Comment:
What is the different between T_0 and the root T? Is oftype(T, p) change the type of p the same as T?

Reply:
`T` is te solution, and `T_0` is the initial guess.

> Is oftype(T, p) change the type of p the same as T?

Basically, yes. `oftype` is used to convert the `Float` type of `p` to a `Dual` for `ForwardDiff`.